충남대 김규현 Step1 엔티티 매핑

HappyGogildong opened the pull request:
Wish : Product = N : 1
Wish : Member = N : 1 로 생각하여 Join을 하게 했습니다.
Fetch 시점은 둘 다 Lazy로 했습니다.

HappyGogildong committed:
<build>(project): 0단계 코드 준비

HappyGogildong committed:
<refactor>(Entity): Jpa 기반으로 Entity 클래스 구조 변경

HappyGogildong committed:
<refactor>(Repository): Jpa 기반으로 repository 인터페이스 구조 변경

HappyGogildong committed:
<refactor>(ProductService, JwtTokenHandler): Jpa 기반으로 ProductServiceImpl 구조 변경

containsProduct 메서드는 Jpa의 existsById로 대체
Member가 record에서 class로 변경되어 Member필드에 접근은 getter 메서드 사용

HappyGogildong committed:
<refactor>(MemberService): Jpa 기반으로 MemberService 구조 변경

User -> Member로 도메인 이름을 변경해 관련된 이름들 다 변경하는 중

HappyGogildong committed:
<refactor>(WishService): Jpa 기반으로 WishService 구조 변경

HappyGogildong committed:
<refactor>(Repository): @Repository 어노테이션 붙임

HappyGogildong committed:
<feat>(Entity Test): Entity 테스트 코드 작성

HappyGogildong committed:
<feat>(Entity Test): Entity 테스트 코드 작성

HappyGogildong committed:
Merge remote-tracking branch 'origin/step1' into step1

HappyGogildong committed:
<fix>(dto): 상품명 제약 조건 수정

최소 상품 수량 0 -> 1

HappyGogildong committed:
Merge branch 'happygogildong' into step1

urarik commented on src/main/java/gift/entity/Member.java:
```diff
@@ -0,0 +1,59 @@
+
+    @Column(nullable = false)
+    @Pattern(regexp = "^(user|manager)$", message = "역할은 'user' 또는 'manager'만 가능합니다.")
+    private String memberRole;
```
enum을 써보는건 어떨까요?

HappyGogildong replied:
MemberRole Enum 클래스를 만들었습니다. valueOf 메서드를 사용해 request로 받은 String을 enum 타입으로 변경해서 사용했습니다.

urarik commented on src/main/java/gift/service/MemberServiceImpl.java:
```diff
@@ -0,0 +1,66 @@
+        this.jwtTokenHandler = jwtTokenHandler;
+    }
+
+    public Member memberWithEncodedPassword(RegisterRequestDto registerRequestDto) {
```
getMember... 로 바꾸는 것이 더 좋을 것 같아요~

HappyGogildong replied:
getMemberWithEncodedPassword로 변경했습니다.
getEncodedMember, getEncodedPassword 의 메서드명도 생각해보았습니다
Member 엔티티에 (Member -> Member) encodePassword 메서드를 넣어 member.encodePassword로 사용하는 방법은 어떻게 생각하시는지 궁금합니다. DB에 들어가는 모든 Member는 비밀번호 암호화가 되어야 하니, 도메인 정책이라고 생각해 도메인에 넣어도 되지 않을까 하고 생각했습니다

urarik replied:
@HappyGogildong 좋은 생각입니다. 다만 암호화에 대한 로직이 Member에 너무 많이 들어가면 더 중요한 도메인 로직들이 가려질 수 있어요. 비밀번호를 암호화하는 것은 도메인 로직일지라도 암호화의 구체적인 부분까지 Member가 책임을 갖는건 조금 과해보이기도 하네요~

urarik commented on src/main/java/gift/service/ProductServiceImpl.java:
```diff
@@ -42,48 +42,41 @@ public List<Product> getAllProducts() {
-            productUpdateRequestDto.price(),
-            productUpdateRequestDto.imageURL());
-        productRepository.updateProduct(product);
+        Product product = productRepository.findById(productId).orElseThrow();
```
해당 클래스의 getProduct를 사용하면 orElseThrow같은 로직의 중복을 줄일 수 있을 것 같아요~

HappyGogildong replied:
getProduct를 사용해서 중복을 없앴습니다

urarik commented on src/main/java/gift/service/ProductServiceImpl.java:
```diff
@@ -24,15 +25,14 @@ public ProductServiceImpl(ProductRepository productRespository) {
     }
 
-
     public Product getProduct(long productId) {
```
repository에선 find고 service에선 get으로 네이밍해주셨는데 이유가 있을까요~?

HappyGogildong replied:
저는 get은 존재함을 가정한 탐색, find는 존재할 수도 있고 없을 수 도 있는 자원을 탐색하는 동작을 표현하고자 했습니다.
그리고 id는 Auto Increment이므로 DB에 들어간 상품만 id가 존재합니다. 그래서 id를 인자로 받아 탐색하는 메서드는 존재함을 가정한 동작을 해야한다고 생각했습니다. 그 외에 productName, price 등의 다른 조건으로 탐색할 경우엔 없을 수 도 있으니 find를 사용해 메서드명을 짓고자 했습니다

urarik commented on src/main/java/gift/service/WishServiceImpl.java:
```diff
@@ -4,64 +4,96 @@
-    private final WishJdbcRepository wishJdbcRepository;
+    private final MemberRepository memberRepository;
+    private final WishRepository wishRepository;
+    private final ProductRepository productRepository;
```
다른 도메인의 엔티티를 가져올 땐 repository로 직접 가져오기보단 service를 거쳐서 가져오는 것이 좋은 습관입니다. 왜 그럴까요~?

HappyGogildong replied:
wishService에서 product, member 도메인에 대한 조작을 하면 책임 분리가 모호해질 수 있을 것 같습니다.

getMemberIdByEmail과 같은 메서드도 memberService로 옮기고, service를 통해 다른 도메인에 접근하도록 했습니다.

urarik commented on src/main/resources/application.properties:
```diff
@@ -1,7 +1,15 @@
-spring.datasource.url=jdbc:h2:mem:test
+#db url, MySQL Dialect
+spring.datasource.url=jdbc:h2:mem:test;
+spring.jpa.properties.hibernate.dialect=org.hibernate.dialect.H2Dialect
```
dialect 가 뭘까요?

HappyGogildong replied:
각 DBMS가 갖는 독자적인 문법, 기능입니다. 
SQL문을 각 DBMS에 맞게 작성해야하고, 그 설정을 간편하게 변경할 수 있게 Jpa 구현체가 제공합니다.

urarik reviewed (changes requested):
코드 깔끔하게 잘 작성해주셨네요~ 바로 머지해도 될 것 같은 퀄리티이지만 코멘트 한 번 확인해주세요~

HappyGogildong committed:
<refactor>(Member): 회원 유형 enum 타입 사용하게 수정

MemberRole 클래스를 entity 패키지에 저장

HappyGogildong committed:
<refactor>(memberpwd encoding method): 회원 비밀번호 암호화 메서드 이름 변경

memberWithEncodedPassword -> getMemberWithEncodedPassword

HappyGogildong committed:
<refactor>(updateProduct): 코드 중복 해결

getProduct에서 상품 존재 검사와 상품 가져오는 동작을 구현했으므로
updateProduct에 해당 로직을 지우고 getProduct로 대체함

HappyGogildong committed:
<refactor>(service): 다른 도메인에 접근은 repository 클래스를 직접 사용하지 않고 service를 통해 접근하도록 변경

wishService의 product, member 도메인 접근 로직 수정
productService, memberService에 메서드 추가

urarik reviewed (approved):
수고 많으셨습니다~ 머지하도록 할게요!
